### ISSUE TITLE
Better `jj split` ergonomics

### ISSUE THREAD

**Issue by samueltardieu** [jj-vcs/jj#1155]
[follow-up to the discussion in #1147 and the issue in #881]

Right now, `jj split [-r rev] .` adds an empty commit between the target commit and its children, while `jj split [-r rev] non-existent-path` inserts an empty commit between the target commit and its parents. This is consistent with the other usages of `jj spilt pathes...` but not really use-friendly, especially the second form.

First of all, we should get a warning if `non-existent-path` does not exist. Should it be limited to `jj split`? Should it be given only if there are no matches at all, or in the more general case should it be given for all patterns given to the matcher that don't resolve into existing files (for every command using the matcher – in this case we should open a new dedicated issue).

Also, should we add an option (or options) to `jj split` to create such an empty commit? Those options could copy the description of the target commit into the non-empty one and let the other description empty, rather than opening two editors.

We could have, for example, `jj split --empty-parent` (short `-p`) and `jj split --empty-child` (short `-c`). What names should be used in this case, and what short options?

**Comment by martinvonz** [comment]
> First of all, we should get a warning if `non-existent-path` does not exist. Should it be limited to `jj split`?

#505 is about about `jj diff`, but we clearly want it to apply to most commands that accept paths.

> We could have, for example, `jj split --empty-parent` (short `-p`) and `jj split --empty-child` (short `-c`). What names should be used in this case, and what short options?

We currently let the parent commit keep the change id. It would be consistent to do that for `jj split --empty-parent` too, but it would probably also not be what the user wants. Maybe we should instead add flags to `jj new` for this? Like `jj new --insert X` to create a new commit on top of X and rebase existing children on top. However, we also need to keep in mind the discussion about disallowing non-head working-copy commits.

**Comment by ilyagr** [comment]
I think warning about nonexistent paths for `split` is technically a bit easier than for `diff`, since there's only one tree to check for nonexistent files.

I don't have a strong opinion right now, but here are some thoughts I do have.

I was thinking it's good enough to have `jj split .`  and `jj split --empty`, since the latter follows the logic of how `jj split` with filenames as arguments works. `jj split --empty` is what `jj split` without arguments would do if it didn't have the interactive diffedit behavior.

For the above, it would be even better to have `jj` support something like git's `:/` and do `jj split :/`.

At the moment, I think that `jj split --empty` should exist regardless of whether we also add extra behaviors to `jj new`. I'm not opposed to `--empty-parent` and `--empty-child` options; I'll see if I can think of better names for them.

**Separate alternative:** I always found it weird that `jj split a` keeps a in the parent commit; my intuition was that I'm asking to "split `a` off". AFAICT, this matches hg's behavior, and I'm not proposing we change that. However, this suggests another possibility: we could have a `jj split --invert` option (also needs a better name) to flip this behavior and then  use `jj split --invert .` instead of `jj split --empty`.

**Comment by ilyagr** [comment]
Another possibility that's closer to Martin's thoughts above is to have `--after/--before` options to `jj split`. The current/default behavior is `--after`. `--before` would keep the change id with the child commit and create a parent commit.

This might make more sense if `jj split a` moved `a` to the "other" commit, in which case `--empty` would still be necessary.

Also, I missed the discussion about disallowing non-head working copies, so I don't account for that. Currently, I prefer head working copies, but I use non-head working copies often simply because so many commands operate on "@" by default.

**Comment by samueltardieu** [comment]
I also agree that "split files" should put files in the "other" commit, as they should be the files that are split off. `--before` could be the default to avoid changing the current behaviour. It would put the selected files/changes into the parent, keeping the change id on the child. `--after` would revert this behaviour and put the selected files/changes into the child, keeping the change id on the parent. In both cases, a branch on the original commit would be set to the child.

If changing the current behaviour is ok, I would prefer `--after` to be the default (with a short version `-A`), and `--before` could have a short option such as `-B`. `-A` and `-B` look like their `grep` counterpart. If we have that we could also exit with an error if a matcher resolving to nothing is given on the command line.

If `--after` or `--before` is used, I would make the operation totally non-interactive and copy the description from the original commit to the new commit containing the files.

**Comment by ilyagr** [comment]
This all sounds reasonable to me except, maybe, for one point.

> If --after or --before is used, I would make the operation totally non-interactive and copy the description from the original commit to the new commit containing the files.

I think there should be an option to change this, e.g. `--no-description` or `--interactive-description` or `--description` with possible values `interactive`, `none`, or the text to put in the other commit's description.

I would lean for the default to stay `interactive`. Partially, that's just habit. Partially, I think I regularly mess up the list of files I give `jj split` on the command line. I then use the interactive description editor (which tells me exactly which files go into the parent commit) as an opportunity to abort the operation.

Ultimately, either way would probably be OK.

**Comment by samueltardieu** [comment]
> I think there should be an option to change this, e.g. --no-description or --interactive-description or --description with possible values interactive, none, or the text to put in the other commit's description.

The more I think about this, the more I am tempted to extend `jj new` for those use-cases instead, with something like `--insert` for `jj split .` and `--insert-before` for `jj split nonexistent` (maybe there are better names). It would do what I want with the description, we would have `-m` to give a description as well. The only change compared to what I would like is that a branch would not be moved to the new commit to stay consistent with `jj new`.

**Comment by samueltardieu** [comment]
I also like the proposals "jj split --empty-parent" and "jj split --empty-child" from #620. They could be the fully non-interactive versions of "jj split nonexistent" and "jj split ." respectively, they wouldn't accept path arguments.

**Comment by samueltardieu** [comment]
I've opened #1167 to experiment with `--empty-parent` and `--empty-child`. Don't hesitate to comment or to voice concerns.

**Comment by samueltardieu** [comment]
> This all sounds reasonable to me except, maybe, for one point.
> 
> > If --after or --before is used, I would make the operation totally non-interactive and copy the description from the original commit to the new commit containing the files.
> 
> I think there should be an option to change this, e.g. `--no-description` or `--interactive-description` or `--description` with possible values `interactive`, `none`, or the text to put in the other commit's description.

I thought about adding a `-m message` option like we have in `jj new` or `jj commit`. But what should it do in interactive mode or when a file name is given? If we restrict its use with `--empty-parent` and `--empty-child`, I am worried about the complexity of the `jj split` cli, even though I would like it for my own use.

**Comment by ilyagr** [comment]
Thank you for thinking about this so carefully! I think many of these approaches will work, and I'm not sure if I have a strong preference at the moment. I'll look at your PR and see if that generates some thoughts.

In particular, I'm happy with adding options to either of `jj split` or `jj new`. I think Yuya and Martin expressed a preference for the latter, but I'm not sure how strong their opinion is.

One minor thought I did have:

> ... he more I am tempted to extend jj new for those use-cases instead ... something like `--insert` for `jj split .` and `--insert-before` for `jj split nonexistent` (maybe there are better names)... The only change compared to what I would like is that a branch would not be moved to the new child commit to stay consistent with jj new.

IMO, the meaning of the work "insert" is flexible enough that it'd be fine to move the branch with `jj new --insert-after`, if that's the direction we pick.

**Comment by ilyagr** [comment]
As I was looking at #1167, I started wondering whether `jj split` could have `--parent-inherits-id/--child-inherits-id` options. The default would be `--parent-inherits-id` unless `--empty-parent` is specified.

**Comment by martinvonz** [comment]
> i've toyed a bit with the idea of `jj new --rebase-children`, which does a `jj new` then rebase the children of the parent onto the new commit.

Thanks!

> * The new commit can have several parents. Should all descendants of all parents be rebased? I prefer `jj split --empty-child` here (even though I miss `-m` to give a description) since it forces you to have one parent only.

That's a good point.

To illustrate your question, it would look like this if we did rebase descendants of all parents:

```
$ jj log
@ E
o D
| o C
| o B
|/  
o A
$ jj new --insert B D
$ jj log
o E'
| o D'
|/
@ F
|\  
o | C
| o B
|/  
o A
```

I agree that it's not at all clear that that's what the user would want.

There's also the question of what the user would want to do next. They would presumably start working on that commit, but I'm not sure they want to make it the working-copy commit (like `jj edit` the new commit). It might be better to create a new working-copy commit on top of it (like `jj checkout` the new commit). What do you think?

One option is to create a dedicated `jj insert` command with `--before` and `--after` options. I'm not sure which should be the default, but I'm leaning towards `--before`, since it would then also work on merge commits. How do you feel about having a dedicated command like that?

> * The children can be rebased only after the new commit is created. Using a `DescendantRebaser` here is not possible since the new commit is itself a child of the parent.
> * You cannot use a `DescendantRebaser` to switch from a root-commit parent into another parent. This prevents you from doing a `jj new --rebase-children root`. Note that you can't `jj split --empty-child -r root` for the same reason at this time.

You should be able to work around those limitations by explicitly rebasing the immediate children of the new commit's parents.

**Comment by ilyagr** [comment]
> I agree that it's not at all clear that that's what the user would want.

The boring answer to this particular issue would be to forbid the revset of `jj new` from containing more than one commit when `--insert`  is used.

**Comment by martinvonz** [comment]
> The boring answer to this particular issue would be to forbid the revset of `jj new` from containing more than one commit when `--insert` is used. Your example is pretty compelling.

Right, and that's why I also brought up the question of whether to `edit` the new commit (like `new` would do), or if it's better to have a new command that doesn't do that. Just an idea, maybe a bad one :)

**Comment by ilyagr** [comment]
Yeah, that's also a very good point. I just don't have much to say about it at the moment.

**Comment by samueltardieu** [comment]
> To illustrate your question, it would look like this if we did rebase descendants of all parents:

I was thinking more about the following situation (`--rebase-children` acts like `jj new` in that it inserts the commit after the given changes):

Starting from this tree :

```
@   F
|\
o | E
| o D
|/
| o C
| o B
|/
o A
```

`jj new --rebase-children B D` would create `G` with `B` and `D` as parents and rebase all previous children from `B` and `D` onto `F`, replacing any `B` or `D` parent by `F` and leaving the other parents untouched:

- `C` parent is `G`
- `F` parents are `G` and `E`

as in:

```
o   F'
|\  
| | o C'
| |/  
|/|   
@ |   G
|\ \  
| | o E
o | | D
| |/  
|/|   
| o B
|/  
o A
```

(incidentally we need a better graph placement algorithm, B should have been put on the left side)

Starting again from the original tree, `jj new --insert C F` would create the following situation (`--insert` inserts the new commit before the given changes):

- `G` parents are `B`, `D`, and `E`
- `C` and `F` parent is `G`

as in:

```
o F'
| o C'
|/
@-.   G
|\ \
o | | E
| o | D
|/ /
| o B
|/
o A
```

Does that seem useful? At least it seems powerful and it looks like it would do the right thing.

I used the fictitious option names `--rebase-children` and `--insert`, but that could something like `--insert-after` and `--insert-before` or things like that.

**Comment by samueltardieu** [comment]
I've opened #1171 to play with this idea.

**Comment by martinvonz** [comment]
> I was thinking more about the following situation (`--rebase-children` acts like `jj new` in that it inserts the commit after the given changes):

That looks what I was also thinking about, but I accidentally swapped C and D in my example 😛. I've updated my comment above.

> (incidentally we need a better graph placement algorithm, B should have been put on the left side)

Yes, I think that's #242 for you're asking for.

> Does that seem useful? At least it seems powerful and it looks like it could do the right thing.

"Powerful but maybe not useful" was exactly what I felt when I wrote my comment :)

I suppose the "unnecessary power" is fine - users are unlikely to pass multiple arguments and expect something else to happen, so they'll just not notice it.

I'm less sure about the edit-vs-checkout behavior. After inserting a new commit, what do you plan to do next? The simplest behavior would be for the command to not change your current working-copy commit. Then you could do something like `jj new --insert @-- -m "update readme"; jj move --to @--- README.md` to effectively split your current working-copy commit and move just the changes to the README into the new commit. That would be different behavior compared to `new` without `--insert`, however, which seems like an indication that this should be a different command. Or maybe users typically do want the command to edit the new working-copy commit?

> Is that expected that the content of the children tree do not change even though they may end up with different ancestors, or should we rebase them instead?

I think we should definitely rebase, not reparent. There's `rewrite::rebase_commit()` for that.

**Comment by samueltardieu** [comment]
> I'm less sure about the edit-vs-checkout behavior. After inserting a new commit, what do you plan to do next? The simplest behavior would be for the command to not change your current working-copy commit. Then you could do something like `jj new --insert @-- -m "update readme"; jj move --to @--- README.md` to effectively split your current working-copy commit and move just the changes to the README into the new commit. That would be different behavior compared to `new` without `--insert`, however, which seems like an indication that this should be a different command. Or maybe users typically do want the command to edit the new working-copy commit?

I wouldn't know. I know that sometimes  I don't like the implied `jj edit`:after a `jj merge` with no conficts, I would prefer to either checkout the merge commit in order to build on top of it (instead of modifying the merge commit itsef), or not change my working copy at all (fortunately, a simple `jj new --insert` or `jj split --empty-parent` would easily fix the situation).  Even with a conflicting `jj merge`, being in a child commit easily lets you `jj squash` the resolution.

Would the new command you have in mind be like the `new` we are discussing, but without the final edit (and `new` would stay as it is now)? Would it be feasible to let `jj merge` not have the final edit too, knowing that `jj new` can be used for that?

> I think we should definitely rebase, not reparent. There's `rewrite::rebase_commit()` for that.

This is what I have in the #1172 playground.

**Comment by martinvonz** [comment]
> I wouldn't know. I know that sometimes I don't like the implied `jj edit`:after a `jj merge` with no conficts, I would prefer to either checkout the merge commit in order to build on top of it (instead of modifying the merge commit itsef), or not change my working copy at all (fortunately, a simple `jj new --insert` or `jj split --empty-parent` would easily fix the situation). Even with a conflicting `jj merge`, being in a child commit easily lets you `jj squash` the resolution.

Interesting. Do you mean that you typically leave the merge commit empty and without description? I rarely create merge commits, but when I do, I typically just let them be what Git calls "evil".

> Would the new command you have in mind be like the `new` we are discussing, but without the final edit (and `new` would stay as it is now)?

Yes, but it would probably also not support multiple parents, at least not until we see a need for it. Again, it's just an idea, and I'm not at all attached to it.

> Would it be feasible to let `jj merge` not have the final edit too, knowing that `jj new` can be used for that?

I think the `jj merge` command is there more for familiarity for users coming from git or hg (i.e. pretty much everyone). Git does create a commit, but hg doesn't. Since most users (or most potential users anyway) come from git, it sounds reasonable if we were to do the same (ask for a description and create a new commit on top). Let's hear what others think.

> > I think we should definitely rebase, not reparent. There's `rewrite::rebase_commit()` for that.
> 
> This is what I have in the #1172 playground.

Ah, nice. Thanks again for working on this! I'll review once we've decided how we think it should work.

**Comment by khionu** [comment]
A bit of a tangent from the above, but also on topic for the issue title, I was thinking maybe there could be a way to drop one of the two new changes by setting the description. Would need to be something super explicit, maybe `JJ-DO: ABANDON`, in spirit with `JJ:` being special.

My current use case could have been avoided with a smidge more foresight, but it would be nice as an option for correcting mistakes?

**Comment by bnjmnt4n** [comment]
@khionu I think that usecase would be better covered by something like `jj diffedit` since it doesn't really align with the idea of splitting commits.
